### ISSUE TITLE
GPU Fitting (excessive transfers)

### ISSUE THREAD

**Issue by yuvi-m** [MUSTANG-SZ/WITCH#230]
Follow up from GPU transfer discussion in #218 

Ran the gaussian probe (unit_tests/gauss_probe.yaml) on the A6000 with jax.transfer_guard('log') to find where we're moving data between host and device.

921 transfers total for a 20 step run. GPU is ~2x faster per step than CPU (1.35 vs 2.76 s), but the transfers are most likely capping the speedup.

Most common transfers (timestamps stripped, counted by type/shape):

497    host-to-device   (small, no shape)
51      device-to-host    shape= ()    F64
42      host-to-device    int64[] weak_type
36      device-to-host    shape= (1)  F64
25      host-to-device    float64[] weak_type
24      device-to-host    shape= (4)  F64
~150  device-to-host    shape= (N, 2362)  F64

I think that two patterns stand out:

1. weak typed scalars going from host to device (497, 42, 25). They look like python ints/floats getting shipped to the device over and over again. 
2. the TOD arrays round-tripping to the host (the (N, 2362) device to host ones). Each objective eval seems to pull the per-TOD data back to the host. 

I'm guessing that the weak-typed scalar transfers are the quick win and the TOD round-trips are the bigger, structural one. I'm happy to dig into the specific call sites; I saved the full log.

You can reproduce it with:

```
import sys, jax
sys.argv = ['witcher', 'unit_tests/gauss_probe.yaml']
from witch.fitter import main
with jax.transfer_guard('log'):
    main()
```

**Comment by Sulla2012** [comment]
I suspect the TOD transfers are killing us. The actual TODs should be able to fit on memory just fine; using MOO1142 as an example, one TOD is ~3.3Mb, about 100 TODs for >~400Mb. What I'm not sure of is when/how the TODs are actually being shuttled too and from the GPU. The data management is opaque even to me. The TODs live in `datasets[0].datavec.tods` (note if there's more than one `dataset` they live in `0,1,2...`), and `datasets` itself is packaged onto `metamodel`. I haven't been able to figure out what the size of `metamodel` itself is. The `fit_loop` is a bunch of bookkeeping and then a call to either `_run_fit` or `_run_mcmc`, which in turn call `run_lmfit` and `run_mcmc`, respectively. These functions each take `metamodel` and then a bunch of integers as arguments. So I'm not sure here when/why `metamodel` is coming onto/off the GPU, but I would think we could just put the entire `metamodel` on GPU and leave it there as it's fairly small. What is big are the compiled functions, but it doesn't seem to me that they're being passed to/from the GPU.

**Comment by yuvi-m** [comment]
I checked where the data actually lives after loading, and the TODs and noise model are already on the GPU: 

```
TOD data (datavec[0].data): CudaDevice(id=0)
noise.v:                 CudaDevice(id=0)
noise.filt_spectrum:           CudaDevice(id=0)
```

So they're not sitting on host and getting pushed every eval; they load onto the device and stay there. That means the (n, 2362) device to host transfers I logged are pulling TOD-sized arrays that are already on the GPU back to host mid-fit, which lines up with your suspicion that something is shuttling them off rather than them not being on the GPU in the first place.

My suspicion is the MPI collectives in `joint_objective`. If mpi4jax isn't using a CUDA-aware MPI, every collective has to bring the array to host, hand it to MPI, and push the result back, which would explain device to host round trips even though the data lives on the GPU. That would also mean your CUDA-aware MPI build is probably the actual fix for the big transfers, not just puts. The gradient allreduce is small ((npar,)) though, so we should probably pin down exactly which array is making the (n, 2362) trip. I'll keep digging with transfer guard tracebacks to find the call site.

**Comment by yuvi-m** [comment]
Following up with a correction to my earlier message; I dug into whether the TOD transfers actually scale with the chain, and they don't, so I don't think they're the per-step bottleneck.

I counted the (n, 2362) device to host transfers after "Running MCMC" at two settings:

1. 20 steps, max_tries=0 ~69 TOD transfers
2. 50 steps, max_tries=3 (5 hmc calls) ~40 TOD transfers

If the TODs were being pulled to host every step, going to more steps and more hmc calls would balloon that count. It stays flat, so the big TOD transfers are happening at compile/first eval time, not per step I believe. The sample function compiles once and the cache holds, so the TOD-sized transfers don't repeat each step.

However, I think that what does look per step is the small stuff; (4,) F64 and (3,)/() PRED (param vector and booleans, respectively), getting pulled to host, which looks like the tuning/acceptance logic reading values back for python control flow each iter.

So, I think that the TOD transfers are real but one time, not the per step cost I thought they were earlier. It might be better to target small scalar/predicate pulls for the puts.
